apps/fz_http/lib/fz_http_web/channels/gateway_channel.ex

```diff
@@ -0,0 +1,55 @@
+defmodule FzHttpWeb.GatewayChannel do
+  @moduledoc """
+  Error handler for halting pipe processing when erroring out when communicating with the gateway
+  """
+  alias FzHttp.Gateways
+
+  use FzHttpWeb, :channel
+
+  require Logger
+
+  @impl Phoenix.Channel
+  def join("gateway:all", _payload, socket) do
+    {:ok, socket}
+  end
+
+  @impl Phoenix.Channel
+  def join("gateway:" <> id, payload, socket) do
```

Comment:
@jamilbk @conectado this is the place we were discussing on the call. Right now to start a gateway it first goes to GatewayController and sends an HTTP(S) request to register itself (create a record in the `gateways` table), and then it gets a JWT with an id of this record back. The Gateway table assumes a name to be present and it causes issues I've described in [another comment](https://github.com/firezone/firezone/pull/1205#discussion_r1065063148).

What I suggest to change:
1. Gateway generates a unique name for itself, it can be something as simple as UUIDv4, its own IP address, a name from an environment variable, FQDN, or a combination of those. I would recommend going with the following: `GATEWAY_ID` environment variable or `hash(public_key <> ip_address)` of the gateway if ENV is not set, or `hash(public_key <> random_bytes())` if we can't get an IP address. 

The gateway should persist the ID in a file, by default it will be ephemeral - re-generated on reboot (but will be the same if we can get the IP), but it will allow some clients to add a volume at the file path and make it persistent if they want to.

3. Then when it boots it doesn't register itself via HTTP(S), instead, it immediately connects to this channel sending the generated ID and `secret` (that it already sends to the controller). The GatewaySocket will use `secret` to authenticate the request the same way as `FzHttpWeb.AuthGatewayController` does now in `check_token` function. Then the GatewayChannel upserts Gateway into DB if it used a new ID and pushes back ALL needed configurations to the Gateway.

The gateway should store received configs in some files too for the same reason as above, ephemeral by default, persistent for DevOps that know what they are doing. It would be nice if the gateway can boot from a such file if the Panel is temporarily not available for some `CONFIGURATION_GRACE_PERIOD_SECONDS` timeout. To make it work we should also keep the timestamp updated in that file every time we lose a WebSocket connection with Panel.

5. Along with connection requests it supplies User-Agent (its version) and peer data that are then persisted to DB in `handle_join`, I bet we want to render those in UI.

6. (Optional.) We might want to add `Phoenix.Presence` here and track all gateways that are currently connected to the socket, so that in UI we distinguish between active ones.

So all env vars gateway can have:
- GATEWAY_SECRET (required) - a random string that will allow the gateway to be authorized with a panel;
- PANEL_ENDPOINT - URL to WebSocket endpoint that the gateway should connect to authenticate, later once we have a cloud panel it will become optional and the default value will point to it;
- GATEWAY_ID (optional) - a name override;
- KEY_PATH (optional) - path to public and private key, or two envs (`PUBLIC_KEY_PATH` and `PRIVATE_KEY_PATH`) if we want to keep them in separate files. We can generate key pair during the boot if env is not set and store it to some well-known path;

Why it's better:
1. We already use secret-based authentication in `FzHttpWeb.AuthGatewayController`, which is exchanged to JWT. JWT doesn't make it more secure, it's just an extra state that we don't need. And it doesn't increase our security in any way but makes it worse by increasing the attack surface (supply chain attacks, for example).
2. We don't need to add `:guardian_phoenix` dependency just for this token exchange;
3. We have way fewer moving pieces and code, everything related to the gateway is in a single place, and the gateway doesn't need to talk HTTPS to the panel at all.

Reply:
How a key rotation would look like: you just green/blue rollout a set of gateways without a private key at a given path (or with a new key you supplied). Old gateways are getting removed and new ones are being added. The gateway name will change by default but will stay persistent if you set it in the ENV variable or attached a volume.

Reply:
So there's one thing I didn't clarify on the call -- the way this works here in this PR was meant to be an interim stepping stone on the path to Multi-site. Only the first gateway (the `default` gateway in this PR) will auto-register using a shared secret on boot. This is doable today because the shared secret can simply be an ENV var that gets generated by the install script and added to `$HOME/.firezone/.env` -- for 0.8 the gateway and portal will still be on the same host. 

We didn't intend for Gateways to auto-register themselves in the longterm. Only the default gateway as part of the install / upgrade process. Instead we had decided the flow will be portal-driven to avoid the possibility someone could spam the portal with gateway registration requests.

I think the goal for 0.8 was to get the data plane piece split out from the portal, with as little breakage as possible for existing users. Then assess whether we want to move forward with multi-site *before* client apps are complete, or wait until client apps are complete before building multi-site.

Reply:
I'm not sure if this furthers the discussion of JWT vs no JWT, but hopefully adds context. The gateway requires Linux with a recent kernel, but otherwise is meant to be lean and flexible enough to be easy to deploy and manage it in a variety of ways (baremetal, Docker, k8s, etc) on a variety of architectures, so that was the thinking behind limiting what's persisted.

A gateway could be "re-registered" if it ever loses its auth token (whether JWT or not) -- would prevent an admin from having to reconfigure a new gateway in case the auth token is ever accidentally lost.

Reply:
I'm not married to JWTs if it's not required, it just provides a convenient and familiar vehicle for accomplishing HTTP auth.

> What I suggest to change:
> 
> 1. Gateway generates a unique name for itself, it can be something as simple as UUIDv4, its own IP address, a name from an environment variable, FQDN, or a combination of those. I would recommend going with the following: `GATEWAY_ID` environment variable or `hash(secret <> public_key <> ip_address)` of the gateway if ENV is not set, or `hash(secret <> public_key <> random_bytes())` if we can't get an IP address.
> 

Whichever auth token format we use, just bear in mind we'll need some sort of key rotation of the WireGuard keys for the gateway at some point. It's not technically possible to do this without flapping the WireGuard interface, but we could minimize the downtime to a few seconds every week or so.

We'd probably also want to rotate the auth token at some interval too, though I'm not sure if they should both be rotated at the same time.

> 4. We have way fewer moving pieces and code, everything related to the gateway is in a single place, and the gateway doesn't need to talk HTTPS to the panel at all.

Hmmm, I'm not sure I'm following. The gateway will talk to the panel over HTTPS because it has no WireGuard tunnel to the portal.

Reply:
Last note on persistence -- maybe it's best to allow for all settings to be configured via ENV vars **or** config file using a convention? With the ENV var taking precedence.

E.g.

| env var | config file |
| --- | --- |
| `FZ_PRIVATE_KEY` | `private_key` |
| `FZ_AUTH_TOKEN` | `auth_token` |
| ... | ... |

Reply:
Is there a reason why we don't want auto-registration? If we have secret auth (like we are doing now) there is no point to pre-register gateways, it won't make Firezone more secure. If we then decide to add some per-gateway restrictions then we can apply some secure default to a newly added gateway (eg. pause it) and wait for the administrator to approve the configuration change, but we should do it later.

In any way, I still suggest applying the proposed changes, it will be less code now and will get us closer to multi-site. Technically, we will have it right away (just not advanced configuration features) and we will be also able to experiment with HA deployments and horizontal scalability.

Reply:
> I'm not married to JWTs if it's not required, it just provides a convenient and familiar vehicle for accomplishing HTTP auth.

Right now we do it two times, once with secret auth and the second with JWT, and have a secret-to-JWT exchange in the middle. I don't think it adds any convenience at all, just makes things more complicated.

> Whichever auth token format we use, just bear in mind we'll need some sort of key rotation of the WireGuard keys for the gateway at some point. It's not technically possible to do this without flapping the WireGuard interface, but we could minimize the downtime to a few seconds every week or so.

> We'd probably also want to rotate the auth token at some interval too, though I'm not sure if they should both be rotated at the same time.

I've explained the rotation procedure "for now", we can make it more graceful too if we want too in the future. Eg. for `secret` we can store a list of secrets, so to rotate you add a new one to the panel, then rollout it to gateways, then remove the old one from the panel.

Reply:
> Hmmm, I'm not sure I'm following. The gateway will talk to the panel over HTTPS because it has no WireGuard tunnel to the portal.

Gateway will use WS(S) protocol for WebSocket (also an HTTP-based protocol), but it won't need to do any REST API calls (and hence have a simple REST API client to the Panel).

Reply:
> Last note on persistence -- maybe it's best to allow for all settings to be configured via ENV vars or config file using a convention? With the ENV var taking precedence.

We should have a naming convention like that but I would not put everything in one file, because then you can't make things persistent granularity. With separate files you can make everything persistent (mounting volume to the folder) or just parts of it (to individual files).

Reply:
> I'm not sure if this furthers the discussion of JWT vs no JWT, but hopefully adds context. The gateway requires Linux with a recent kernel, but otherwise is meant to be lean and flexible enough to be easy to deploy and manage it in a variety of ways (baremetal, Docker, k8s, etc) on a variety of architectures, so that was the thinking behind simplifying the persistence to e.g. a single json file. It could certainly generate its own unique ID and token too, with a flow like:
> 
>     1. Admin creates gateway registration request in portal
> 
>     2. Gateway registers using a one-time-use registration URL (contains the short-lived registration token). Sends its unique ID and UserAgent, etc
> 
>     3. Admin confirms gateway registration in UI -- gateway is now registered
> 
> 
> A gateway could be "re-registered" if it ever loses its auth token (whether JWT or not) -- would prevent an admin from having to reconfigure a new gateway in case the auth token is ever accidentally lost.

@AndrewDryga following on this, I think firezone/product#480 can add a lot of context for this.

Reply:
> Last note on persistence -- maybe it's best to allow for all settings to be configured via ENV vars **or** config file using a convention? With the ENV var taking precedence.
> 
> E.g.
> env var 	config file key
> `FZ_PRIVATE_KEY` 	`private_key`
> `FZ_AUTH_TOKEN` 	`auth_token`
> 
> ```json
> {
>   "private_key": "...",
>   "auth_token": "..."
> }
> ```

We can even leverage [this config crate](https://docs.rs/config/latest/config/) to have rust just use a variety of formats and not only json.

Reply:
> Is there a reason why we don't want auto-registration? If we have secret auth (like we are doing now) there is no point to pre-register gateways, it won't make Firezone more secure. If we then decide to add some per-gateway restrictions like manual registration then we can apply some secure default to a newly added gateway (eg. pause it) and wait for the administrator to approve the configuration change, but we should do it later.
> 
> In any way, I still suggest applying the proposed changes, it will be less code now and will get us closer to multi-site. Technically, we will have it right away (just not advanced configuration features) and we will be also able to experiment with HA deployments and horizontal scalability.

The pre-registration is just a stepping stone to having actual manual registration, the idea is that an admin would manually create the gateway and that would:
1. Create the gateway in the DB
1. Create a short-lived registration token which will then be used for the gateway to get the (for now) JWT

This have the advantage of the `GATEWAY_SECRET`  suggested before to be short-lived, so the only long-lived secret will be exchanged through an https REST call and is less prone to be leaked.

Furthermore, I think the complexity added by adding a paused state to the gateway while it waits for a valid config will be more complex than the code necessary to handle the HTTPs register call.

And there should never be multiple gateways using the same registration token, each gateway has a different one.

Reply:
> The Gateway table assumes a name to be present and it causes issues I've described in https://github.com/firezone/firezone/pull/1205#discussion_r1065063148.

This is fine, since each gateway would require to be configured manually, and if the name is never set it will be auto-generated.

Reply:
> Whichever auth token format we use, just bear in mind we'll need some sort of key rotation of the WireGuard keys for the gateway at some point. It's not technically possible to do this without flapping the WireGuard interface, but we could minimize the downtime to a few seconds every week or so.

I think you don't even need to restart the wireguard interface with `wireguard-control`, and for the clients it will be just changing the peer list. So yeah it'd be just a few seconds.

This has the advantage, against the blue/green deployment that it doesn't need any complex kubernetes or similar setup to have rotation. And we want to be able to provide this key rotation to the simplest of setups where firezone might be running.

Reply:
After a call with @conectado here is what we settled on:
1. In this PR and in Gateway we will remove the `AuthGatewayController` and the extra step of exchange of `secret` to the JWT token, to keep things simple. Gateway will start the WS connection right away and send the `token` there.

2. We don't promote this use case yet, but we will most basic support for HA and scalable setups with 0.8 because `secret` right now is de-facto a multi-use token that never expires. (But the WS endpoint will be only available within the private network so it's not a big deal.) 
- To enable HA/autoscaler will need to, ** for now**, share the `secret`, `private_key`, and `public_key` effectively resolving to the same record in the `gateways` table. This is because without own client we can't put multiple public keys to the client config.
- We acknowledge that sharing `private_key` is not a long-term solution, this is an option just so that we can run experiments with HA deployments to k8s, scaling, and make sure other load balancers don't suffer from the same Wireguard compatibility issue like AWS one does. For internal use and to pivot with some of the clients that are willing to take the risk.
- HA mode will be only semi-operational, it won't have an automatic failover on the client. So if the gateway node is down you will need to turn off and then turn back on the WG client connection (restarting the virtual interface) so that a new session is established with a new Gateway replica.

3. Long-term we want to have short-lived tokens and probably multi-use ones for autoscaled/HA setups if they proved to be commonly requested by our users. So multi-use tokens are up for further discussion.

4. We will also need to discuss the UX for (3) once we support multi-site or multi-gateway setups officially.

cc @jamilbk